src/mawpy/steps/trace_segmentation_clustering.py

```diff
@@ -0,0 +1,145 @@
+
+    """
+        For the trace of a user on a given day, the function calculates and assigns stay_lat and stay_long
+        to each of the daily trace.
+
+        It groups together traces into a 'stay' for which time difference between
+            all the calculated diameter is within the spatial constraint and
+            the first and the last point exceeds duration_threshold
+    """
+    latitudes_for_day = each_day_df[ORIG_LAT].to_numpy()
+    longitudes_for_day = each_day_df[ORIG_LONG].to_numpy()
+    timestamps_for_day = each_day_df[UNIX_START_T].to_numpy()
+    number_of_traces_for_day = len(each_day_df)
+    stay_lat = np.full(number_of_traces_for_day, -1.0)
+    stay_long = np.full(number_of_traces_for_day, -1.0)
+    stay_dur = np.full(number_of_traces_for_day, -1)
+
+    start = 0
+    end = start + 1
+    group_found = False
+    while start < end < number_of_traces_for_day:
+        has_exceeded, exceed_index = _get_diameter_constraint_exceed_index(start, end, latitudes_for_day, longitudes_for_day,
+                                                             spatial_constraint)
+        if group_found is False:
+
+            if has_exceeded:
+                start = exceed_index + 1
+                if start == end:
+                    end += 1
+                continue
+
+            if _does_duration_threshold_exceed(start, end, timestamps_for_day, dur_constraint):
+                group_found = True
+            end += 1
+        else:
+
+            if has_exceeded:
+                stay_lat[start: end] = np.mean(latitudes_for_day[start: end])
+                stay_long[start: end] = np.mean(longitudes_for_day[start: end])
+                stay_dur[start: end] = timestamps_for_day[end - 1] - timestamps_for_day[start]
+                start = end
+                end = start + 1
+                group_found = False
+            else:
+                end += 1
+
+    each_day_df[STAY_LAT] = stay_lat
+    each_day_df[STAY_LONG] = stay_long
+    each_day_df[STAY_DUR] = stay_dur
+
+    return each_day_df
+
+
+def _run_for_user(df_by_user: pd.DataFrame, spatial_constraint: float, dur_constraint: float) -> pd.DataFrame:
+    df_with_stay = df_by_user.groupby(UNIX_START_DATE).apply(lambda x: _get_df_with_stays(x, spatial_constraint, dur_constraint))
+    df_with_stay[STAY] = get_stay_groups(df_with_stay)
+
+    df_with_stay_added = get_combined_stay(df_with_stay)
+
+    return df_with_stay_added
+
+
+def _run(args: tuple) -> pd.DataFrame:
+    df_by_user_chunk, spatial_constraint, dur_constraint = args
+    df_by_user_chunk = (df_by_user_chunk.groupby(USER_ID)
+                        .apply(lambda x: _run_for_user(x, spatial_constraint, dur_constraint)))
+    return df_by_user_chunk
+
+
+def trace_segmentation_clustering(input_file: str, output_file: str, spatial_constraint: float,
+                                  dur_constraint: float) -> pd.DataFrame:
+
+    pool = Pool(cpu_count())
```

Comment:
cpu_count() is the default, so no need to specify here

Reply:
Thanks! updated.